include/fir/Type.h

```diff
@@ -0,0 +1,280 @@
+                     detail::FIRDimsTypeStorage> {
+public:
+  using Base::Base;
+  static DimsType get(mlir::MLIRContext *ctx, unsigned rank);
+  static bool kindof(unsigned kind) { return kind == FIRTypeKind::FIR_DIMS; }
+
+  /// returns -1 if the rank is unknown
+  int getRank() const;
+};
+
+class FieldType : public mlir::Type::TypeBase<FieldType, mlir::Type,
+                      detail::FIRFieldTypeStorage> {
+public:
+  using Base::Base;
+  static FieldType get(mlir::MLIRContext *ctxt, KindTy _ = 0);
+  static bool kindof(unsigned kind) { return kind == FIRTypeKind::FIR_FIELD; }
+};
+
+class HeapType : public mlir::Type::TypeBase<HeapType, mlir::Type,
+                     detail::FIRHeapTypeStorage> {
+public:
+  using Base::Base;
+  static HeapType get(mlir::Type elementType);
+  static bool kindof(unsigned kind) { return kind == FIRTypeKind::FIR_HEAP; }
+
+  mlir::Type getEleTy() const;
+};
+
+class PointerType : public mlir::Type::TypeBase<PointerType, mlir::Type,
+                        detail::FIRPointerTypeStorage> {
+public:
+  using Base::Base;
+  static PointerType get(mlir::Type elementType);
+  static bool kindof(unsigned kind) { return kind == FIRTypeKind::FIR_POINTER; }
+
+  mlir::Type getEleTy() const;
+};
+
+class ReferenceType : public mlir::Type::TypeBase<ReferenceType, mlir::Type,
+                          detail::FIRReferenceTypeStorage> {
+public:
+  using Base::Base;
+  static ReferenceType get(mlir::Type elementType);
+  static bool kindof(unsigned kind) {
+    return kind == FIRTypeKind::FIR_REFERENCE;
+  }
+
+  mlir::Type getEleTy() const;
+};
+
+class SequenceType : public mlir::Type::TypeBase<SequenceType, mlir::Type,
+                         detail::FIRSequenceTypeStorage> {
+public:
+  using Base::Base;
+  struct Unknown {};
+  using BoundInfo = int64_t;
+  using Extent = std::variant<Unknown, BoundInfo>;
+  using Bounds = std::vector<Extent>;
+  using Shape = std::variant<Unknown, Bounds>;
+
+  mlir::Type getEleTy() const;
+  Shape getShape() const;
+
+  static SequenceType get(const Shape &shape, mlir::Type elementType);
+  static bool kindof(unsigned kind) {
+    return kind == FIRTypeKind::FIR_SEQUENCE;
+  }
+};
+
+bool operator==(const SequenceType::Shape &, const SequenceType::Shape &);
+llvm::hash_code hash_value(const SequenceType::Extent &);
+llvm::hash_code hash_value(const SequenceType::Shape &);
+
+
+class TypeDescType : public mlir::Type::TypeBase<TypeDescType, mlir::Type,
+                         detail::FIRTypeDescTypeStorage> {
+public:
+  using Base::Base;
+  static TypeDescType get(mlir::Type ofType);
+  static bool kindof(unsigned kind) {
+    return kind == FIRTypeKind::FIR_TYPEDESC;
+  }
+  mlir::Type getOfTy() const;
+};
+
+// Derived types
+
+class RecordType : public mlir::Type::TypeBase<RecordType, mlir::Type,
+                       detail::FIRRecordTypeStorage> {
+public:
+  using Base::Base;
+  using TypePair = std::pair<std::string, mlir::Type>;
```

Comment:
When using std types, we prefer to include the headers that define those types even if they are implicitly included in other headers.

Reply:
https://llvm.org/docs/CodingStandards.html#minimal-list-of-includes

Reply:
I believe the "minimum" is advising you not to include things you aren't using because the following sentence is: "You must include all of the header files that you are using — you can include them either directly or indirectly through another header file."  And we prefer that such header files are included directly.

Reply:
I'm not going to add unnecessary extra headers to this include file.